File: python-client/p2pshare/protocol/handler.py
"""Message dispatch — routes decrypted messages to the appropriate handler."""

import math

from p2pshare.errors import ProtocolError
from p2pshare.protocol import messages


class MessageHandler:
    """Routes incoming messages to registered callbacks.

    Validates the message envelope (type, version, seq) before dispatching.
    """

    def __init__(self):
        self._handlers: dict[str, callable] = {}
        self._expected_seq = 1

    def register(self, msg_type: str, handler: callable) -> None:
        self._handlers[msg_type] = handler

    def handle(self, msg: dict) -> dict | None:
        """Process an incoming decrypted message. Returns a response dict or None."""
        # Validate envelope
        msg_type = msg.get("type")
        if not msg_type:
            raise ProtocolError("Message missing 'type' field")

        version = msg.get("version")
        if version != messages.PROTOCOL_VERSION:
            return messages.error_msg(0, messages.VERSION_MISMATCH,
                                      f"Unsupported protocol version: {version}")

        handler = self._handlers.get(msg_type)
        if handler is None:
            raise ProtocolError(f"Unknown message type: {msg_type}")

        return handler(msg)

File: go-client/internal/protocol/handler.go
package protocol

import (
	"fmt"
)

// HandlerFunc is the callback type for message handlers.
type HandlerFunc func(msg map[string]interface{}) (map[string]interface{}, error)

// MessageHandler routes incoming messages to registered callbacks.
type MessageHandler struct {
	handlers map[string]HandlerFunc
}

// NewMessageHandler creates a new handler.
func NewMessageHandler() *MessageHandler {
	return &MessageHandler{handlers: make(map[string]HandlerFunc)}
}

// Register registers a handler for a message type.
func (mh *MessageHandler) Register(msgType string, handler HandlerFunc) {
	mh.handlers[msgType] = handler
}

// Handle processes an incoming decrypted message.
func (mh *MessageHandler) Handle(msg map[string]interface{}) (map[string]interface{}, error) {
	msgType, _ := msg["type"].(string)
	if msgType == "" {
		return nil, fmt.Errorf("message missing 'type' field")
	}

	version, _ := msg["version"].(string)
	if version != ProtocolVersion {
		return ErrorMsg(0, VersionMismatch,
			fmt.Sprintf("Unsupported protocol version: %s", version)), nil
	}

	handler, ok := mh.handlers[msgType]
	if !ok {
		return nil, fmt.Errorf("unknown message type: %s", msgType)
	}

	return handler(msg)
}
